src/driver-asus.c

```diff
@@ -189,51 +192,51 @@ static int
 asus_driver_save_profile(struct ratbag_device *device, struct ratbag_profile *profile)
 {
 	int rc = 0;
-	const int8_t *bmi = ratbag_device_data_asus_get_button_mapping(device->data);
-	struct asus_button *asus_button;
 	struct ratbag_button *button;
 	struct ratbag_led *led;
 	struct ratbag_resolution *resolution;
+	struct asus_data *drv_data = ratbag_get_drv_data(device);
 
 	/* set buttons */
 	ratbag_profile_for_each_button(profile, button) {
 		if (!button->dirty)
 			continue;
 
-		int asus_index = bmi[button->index] != -1 ? bmi[button->index] : ASUS_BUTTON_MAPPING_INDEX[button->index];
-		int asus_code;
+		int asus_index = drv_data->button_indices[button->index];
 		if (asus_index == -1) {
 			log_debug(device->ratbag, "No mapping for button %d\n", button->index);
 			continue;
 		}
 
-		log_debug(device->ratbag, "Button %d (%d) changed\n", button->index, asus_index);
+		int asus_code_src = drv_data->button_mapping[asus_index];
+		int asus_code_dst;
+		struct asus_button *asus_button;
+		log_debug(device->ratbag, "Button %d (%02x) changed\n",
+			  button->index, (uint8_t) asus_code_src);
 
 		switch (button->action.type) {
 		case RATBAG_BUTTON_ACTION_TYPE_NONE:
 			rc = asus_set_button_action(
-				device, asus_index,
-				ASUS_BUTTON_CODE_DISABLED,
+				device, asus_code_src, ASUS_BUTTON_CODE_DISABLED,
 				ASUS_BUTTON_ACTION_TYPE_BUTTON);
 			break;
 
 		case RATBAG_BUTTON_ACTION_TYPE_KEY:
 			/* Linux code to ASUS code */
-			asus_code = asus_find_key_code(button->action.action.key.key);
-			if (asus_code >= 0)
+			asus_code_dst = asus_find_key_code(button->action.action.key.key);
```

Comment:
Same thing as with `asus_code_src` - the sentinel `-1` is not handled.
Also, because of this a few lines below there is an implicit conversion from int to u8.